main.nf

```diff
@@ -0,0 +1,581 @@
+if(workflow.profile == 'awsbatch'){
+   summary['AWS Region'] = params.awsregion
+   summary['AWS Queue'] = params.awsqueue
+}
+if(params.email) summary['E-mail Address'] = params.email
+log.info summary.collect { k,v -> "${k.padRight(15)}: $v" }.join("\n")
+log.info "========================================="
+
+
+def create_workflow_summary(summary) {
+
+    def yaml_file = workDir.resolve('workflow_summary_mqc.yaml')
+    yaml_file.text  = """
+    id: 'nf-core-deepvariant-summary'
+    description: " - this information is collected when the pipeline is started."
+    section_name: 'nf-core/deepvariant Workflow Summary'
+    section_href: 'https://github.com/nf-core/deepvariant'
+    plot_type: 'html'
+    data: |
+        <dl class=\"dl-horizontal\">
+${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
+        </dl>
+    """.stripIndent()
+
+   return yaml_file
+}
+
+
+/********************************************************************
+  process preprocessFASTA
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+
+process preprocessFASTA{
+
+  publishDir "$baseDir/sampleDerivatives"
+
+
+  input:
+  file fasta from fasta
+  file fai from fai
+  file fastagz from fastagz
+  file gzfai from gzfai
+  file gzi from gzi
+  output:
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi") into fastaChannel
+  script:
+  """
+  [[ "${params.fai}"=="nofai" ]] &&  samtools faidx $fasta || echo " fai file of user is used, not created"
+  [[ "${params.fastagz}"=="nofastagz" ]]  && bgzip -c ${fasta} > ${fasta}.gz || echo "fasta.gz file of user is used, not created "
+  [[ "${params.gzi}"=="nogzi" ]] && bgzip -c -i ${fasta} > ${fasta}.gz || echo "gzi file of user is used, not created"
+  [[ "${params.gzfai}"=="nogzfai" ]] && samtools faidx "${fasta}.gz" || echo "gz.fai file of user is used, not created"
+  """
+
+}
+
+
+/********************************************************************
+  process preprocessBAM
+  If the user gives the index files for the bam files as an input, they are used
+  If not they are produced in this process given only the fasta file.
+  Moreover this takes care of the read group line too.
+********************************************************************/
+
+
+process preprocessBAM{
+
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+  script:
+  """
+	  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { java -jar /picard.jar AddOrReplaceReadGroups \
```

Comment:
Same as above with the logic. Better to inspect the given input channels in the nextflow code for required files instead of testing for them with bash in a process.

If using bioconda Picard then this should just be `picard` instead of `java -jar /picard.jar`

Reply:
Please ask for help in the gitter channel if you want pointers about how to write logic based on channels - we've done this a few times now and it's not super trivial (but not too bad once you've done it a couple of times).

Reply:
Hey @ewels,

The conditional is based upon `samtools` which is only available within the bash block. So I would need to make a process to save the value of `samtools view -H ${bam[0]} | grep '@RG' | wc -l` to a file/variable and then make another process that is only executed when the when the value is 0. 

Do you think this would be better than how it is currently?
Currently it is like this:
```nextflow
process preprocessBAM {

  tag "${bam[0]}"
  publishDir "$baseDir/sampleDerivatives"

  input:
  set val(prefix), file(bam) from bamChannel

  output:
  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel

  script:
  """
  mkdir ready
  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
  I=${bam[0]} \
  O=ready/${bam[0]} \
  RGID=${params.rgid} \
  RGLB=${params.rglb} \
  RGPL=${params.rgpl} \
  RGPU=${params.rgpu} \
  RGSM=${params.rgsm};}
  cd ready ;samtools index ${bam[0]};
  """
}
```

Thanks in advance

Reply:
Hmm, yeah that makes it more complex, I think I was getting tired by the time I got down this far 😅
Does it matter if you run the Picard command on a file that has an `@RG` header already though? What happens if you do it twice - do you actually need to do that logic?

Reply:
@nf-core/core - any ideas?